syntax = "proto3";
message dm_V2 {
            string id_str      =  1; // [0][15].extra.id_str 新idstr
            uint32 mode        =  2; // [0] [1] 弹幕类型
            uint32 fontsize    =  3; // [0] [2] 弹幕字号
            uint32 color       =  4; // [0] [3] 弹幕颜色
            string mid_hash    =  5; // [0] [7] 发送者midHash
            string content     =  6; // [1]    弹幕正文
            uint64 ctime       =  7; // [0] [4] 发送时间TimeStamp(ms)
    repeated bytes unknown_08  =  8; // [#][##]
             int64 dmid        =  9; // [0] [5] dmid
    repeated bytes unknown_10  = 10; // [#][##]
            uint32 type        = 11; // [0] [9] 
       chat_bubble chat_bubble = 12; // [0][11] chatBubble
            uint32 DmType      = 13; // [0][12] DmType
    repeated emots emots       = 14; // [0][15] 表情 repeated
    repeated bytes unknown_15  = 15; // 
    repeated bytes unknown_16  = 16; // 
           lottery lottery     = 17; // [0][16] 抽奖
    repeated bytes unknown_18  = 18; // 
        validation validation  = 19; // [9]   
          userinfo userinfo    = 20; // [2]    用户主站信息
     fan_medal_ext fan_medal_ext=21; // [3]    粉丝牌
    repeated bytes unknown_22  = 22; // 
    repeated bytes unknown_23  = 23; // 
    repeated bytes unknown_24  = 24; // 
    repeated bytes unknown_25  = 25; // 
    repeated bytes unknown_26  = 26; // 
    repeated bytes unknown_27  = 27; // 
    repeated bytes unknown_28  = 28; // 
    repeated bytes unknown_29  = 29; // 
    repeated bytes unknown_30  = 30; // 
    repeated bytes unknown_31  = 31; // 
    repeated bytes unknown_32  = 32; // 
}
message chat_bubble{//12
    uint32 chat_bubble_type  = 1;               // [0][10]
    string chat_bubble_color = 2;               // [0][11]
    // repeated bytes unknown_chat_bubble_03 = 3;
    // repeated bytes unknown_chat_bubble_04 = 4;
}
message emots{//14
    string description = 1;                     // 
    emots_detail detail = 2;                    // 
    repeated bytes unknown_emots_03 =  3;
    repeated bytes unknown_emots_04 =  4;
}
message emots_detail{//14-2
    string emoticon_unique = 1;                 // [0][15].extra.emots.emoticon_unique
    string emots_url = 2;                       // [0][15].extra.emots.emots_url
    uint32 is_dynamic = 3;                      // [0][15].extra.emots.is_dynamic
    uint32 in_player_area = 4;                  // [0][15].extra.emots.in_player_area
    uint32 bulge_display = 5;                   // [0][15].extra.emots.bulge_display
    uint32 height = 6;                          // [0][15].extra.emots.height
    uint32 width = 7;                           // [0][15].extra.emots.width
    repeated bytes unknown_emots_detail_08 =  8;
}
message lottery{//17
    repeated bytes unknown_lottery_01 = 1;
    uint32 activity_source = 2;                 // [0][16]
    string activity_identity = 3;               // [0][16]
    uint32 not_show = 4;                        // [0][16]
    repeated bytes unknown_lottery_05 =  5;
    repeated bytes unknown_lottery_06 =  6;
    repeated bytes unknown_lottery_07 =  7;
    repeated bytes unknown_lottery_08 =  8;
}
message validation{//19
    string ct = 1;
    uint32 ts = 2;
    // repeated bytes unknown_validation_03 =  3;
    // repeated bytes unknown_validation_04 =  4;
}
message userinfo{//20
    uint64 uid = 1;                             // [2][0] 用户uid
    string username = 2;                        // [2][1] 用户名
    string username_color = 3;                  // [2][7] usernameColor
    string face = 4;                            // 用户头像URL
    repeated bytes unknown_userinfo_05 = 5;     // [2][3]?
    repeated bytes unknown_userinfo_06 = 6;     // [2][4]?
    uint32 rank = 7;                            // [2][5] 实名验证？手机号验证？
    uint32 verify = 8;                          // [2][6]? 始终为1？
    repeated bytes unknown_userinfo_09 = 9;     // 
    uint32 is_admin = 10;                       // [2][2] 房管
    fan_medal fan_medal = 11;                   // [3] 粉丝牌
    live_user_info live_user_info = 12;         // [4] 用户直播区信息
    title title = 13;                           // [5] 头衔
    repeated bytes unknown_userinfo_14 = 14;    // ？string
    repeated bytes unknown_userinfo_15 = 15;    // 
    repeated bytes unknown_userinfo_16 = 16;    // 
}
message fan_medal{//20-11
    uint32 medal_level = 1;                     // [3][0] 粉丝牌等级
    string medal_name = 2;                      // [3][1] 粉丝团称号
    repeated bytes unknown_fan_medal_03 = 3;    // 
    uint32 medal_color = 4;                     // [3][4]or[3][7]
    repeated bytes unknown_fan_medal_05 = 5;    // 
    uint32 medal_color_start = 6;               // [3][8]
    uint32 medal_color_border = 7;              // [3][4]or[3][7]
    uint32 medal_color_end = 8;                 // [3][9]
    uint32 anchor_guard_level = 9;              // [3][10] 粉丝牌舰长等级
    uint32 is_lighted = 10;                     // [3][11]?
    repeated bytes unknown_fan_medal_11 = 11;
    repeated bytes unknown_fan_medal_12 = 12;
    repeated bytes unknown_fan_medal_13 = 13;
    repeated bytes unknown_fan_medal_14 = 14;
    repeated bytes unknown_fan_medal_15 = 15;
    repeated bytes unknown_fan_medal_16 = 16;
}
message live_user_info{//20-12
    uint32 User_Level = 1;                      // [4][0]
    uint32 UL_color = 2;                        // [4][2]
    string live_rank = 3;                       // [4][3]
    repeated bytes unknown_live_user_info_04 = 4;
}
message title{//20-13
    string title1 = 1;                          // [5][x]
    string title2 = 2;                          // [5][x]
    repeated bytes unknown_title_03 =  3;
    repeated bytes unknown_title_04 =  4;
}
message fan_medal_ext{//21
    uint64 anchor_id = 1;                       // [3][12] anchorId
    string anchor_username = 2;                 // [3] [2] anchorUsername
    uint32 roomid = 3;                          // 当前直播间 长/短 ID
    repeated bytes unknown_fan_medal_ext_04 = 4;
}